File: day03/day3.js
const fs = require('fs');
const readFile = (path) => fs.readFileSync(path, { encoding: 'utf8' });


const TreeMap = (file) => {
  const data = readFile(file).split('\n')
  const rowLength = data[0].length;
  const numberOfTrees = (location) => location === '#' ? 1 : 0;

  const countTreesForSlope = (rightStep, downStep) => {
    let xPos = rightStep * -1;
    return data.reduce((treeCount, _, yPos) => {
      if (yPos % downStep === 0) {
        xPos = (xPos + rightStep) % rowLength;
        const location = data[yPos][xPos];
        treeCount += numberOfTrees(location);
      }
      return treeCount;
    }, 0);
  }

  const calculateAllSlopes = () => {
    return [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]
      .reduce((product, coordSet) => product * countTreesForSlope(...coordSet), 1);
  }

  return {
    countTreesForSlope,
    calculateAllSlopes
  }
}

const mapper = TreeMap(process.argv[2]);
mapper.countTreesForSlope(3, 1); // 211
mapper.calculateAllSlopes(); // 3584591857

File: day03/day3.rb
# frozen_string_literal: true

# day 3
class TreeMap
  def initialize(file_input)
    @data = File.readlines(file_input).map(&:chomp)
  end

  def count_trees_for_slope(right_step, down_step)
    x_pos = right_step * -1
    @data.to_enum.with_index.count do |_, y_pos|
      next unless (y_pos % down_step).zero?

      x_pos = (x_pos + right_step) % row_length
      @data[y_pos][x_pos] == '#'
    end
  end

  def calculate_multiple_slopes
    [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]
      .map { |coord_set| count_trees_for_slope(*coord_set) }
      .reduce(:*)
  end

  private

  def row_length
    @data[0].size
  end
end

TreeMap.new(ARGV[0]).count_trees_for_slope(3, 1) # -> 211
TreeMap.new(ARGV[0]).calculate_multiple_slopes # -> 3584591857
